README.md
# :black_square_button: react-native-grid-component

[![NPM version](http://img.shields.io/npm/v/react-native-grid-component.svg?style=flat)](https://www.npmjs.com/package/react-native-grid-component)
[![Build Status](https://travis-ci.org/phil-r/react-native-grid-component.svg?branch=master)](https://travis-ci.org/phil-r/react-native-grid-component)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

Easy to use grid component for your [react-native](https://github.com/facebook/react-native/) project. Supports iOS and Android.

Based on [react-native framework](https://github.com/facebook/react-native/) by Facebook.

## Installation

```
npm install react-native-grid-component
```

or

```
yarn add react-native-grid-component
```

## Examples

- [Simple](https://snack.expo.io/@phil/react-native-grid-component-example)
- [Sections](https://snack.expo.io/@phil/grid-with-sections-example)

## Demo

![android low](https://cloud.githubusercontent.com/assets/577316/18456263/d6b977e2-794f-11e6-878f-5737355111ac.gif) ![ios low](https://cloud.githubusercontent.com/assets/577316/18456262/d6b7a39a-794f-11e6-8587-06757dc42e14.gif)

## Usage

```js
import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';

import Grid from 'react-native-grid-component';

class Simple extends Component {
  _renderItem = (data, i) => (
    <View style={[{ backgroundColor: data }, styles.item]} key={i} />
  );

  _renderPlaceholder = i => <View style={styles.item} key={i} />;

  render() {
    return (
      <Grid
        style={styles.list}
        renderItem={this._renderItem}
        renderPlaceholder={this._renderPlaceholder}
        data={['black', 'white', 'red', 'green', 'blue']}
        numColumns={2}
      />
    );
  }
}

const styles = StyleSheet.create({
  item: {
    flex: 1,
    height: 160,
    margin: 1
  },
  list: {
    flex: 1
  }
});
```

[Full interactive example](https://snack.expo.io/@phil/react-native-grid-component-example)

## Props

| Name                           | propType                                       | description                                                                                                 |
| ------------------------------ | ---------------------------------------------- | ----------------------------------------------------------------------------------------------------------- |
| `data` _required_              | Array                                          | Array that will be used to render items                                                                     |
| `renderItem` _required_        | func `(data: any, i: number) => React$Element` | function that accepts data and returns React Element that will be rendered for each `data` item             |
| `renderPlaceholder` _optional_ | func `(i: number) => React$Element`            | function that returns placeholder React Elements that ere rendered to fill the space at the end of the grid |
| `numColumns` _optional_        | number                                         | number of elements per one row                                                                              |
| `keyExtractor` _optional_      | func `(item: object, index: number) => string` | [see docs](https://facebook.github.io/react-native/docs/flatlist#keyextractor)                              |
| `sections` _optional_          | bool                                           | switches internal implementation to [SectionList](https://facebook.github.io/react-native/docs/sectionlist) |

Also see [FlatList](https://facebook.github.io/react-native/docs/flatlist#props) props or [SectionList](https://facebook.github.io/react-native/docs/sectionlist#props) props(if you set `sections` to true)

## See also

- [React Native Looped Carousel](https://github.com/appintheair/react-native-looped-carousel)
- [React Native Buyscreen](https://github.com/appintheair/react-native-buyscreen)

## ISC License (ISC)

Copyright (c) 2016-2019, Phil Rukin <mailto:<philipp@rukin.me>>

Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby granted, provided that the above copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


CHANGELOG.md
# Changelog

## [2.0.0] - 2019-05-18

- `FlatList` and `SectionList` are used now 🎉
- **Breaking** `renderSectionHeader` now works as in [SectionList](https://facebook.github.io/react-native/docs/sectionlist#rendersectionheader), you **need** to update this or your section titles will disappear
- _Deprecation_ sectioned grid will no longer support `data` as an object in the future, new format is `[{title, data:[]}, ...]`
- _Deprecation_ `itemsPerRow` is now deprecated in favor of [`numColumns`](https://facebook.github.io/react-native/docs/flatlist#numcolumns)
- [`keyExtractor`](https://facebook.github.io/react-native/docs/sectionlist#keyextractor) prop is needed now
- `itemHasChanged` prop is no longer needed

## [1.1.0] - 2018-05-19

### Changed

- Replace `componentWillReceiveProps` with `getDerivedStateFromProps`
- Bump `peerDependencies` for `react` and `react-native`


index.js
/**
 * React Native Grid Component
 * https://github.com/phil-r/react-native-grid-component
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  View,
  Dimensions,
  FlatList,
  SectionList
} from 'react-native';

const { width } = Dimensions.get('window');

// http://stackoverflow.com/questions/8495687/split-array-into-chunks
// I don't see the reason to take lodash.chunk for this
const chunk = (arr, n) =>
  Array.from(Array(Math.ceil(arr.length / n)), (_, i) =>
    arr.slice(i * n, i * n + n)
  );

const prepareData = ({ data, numColumns, itemsPerRow }) => {
  const col = numColumns || itemsPerRow;
  const missing = data.length % col;
  return [
    ...data,
    ...Array.from(Array(col - missing)).map(() => ({ __empty: true }))
  ];
};

const prepareSectionedData = ({ data, numColumns, itemsPerRow }) => {
  const col = numColumns || itemsPerRow;
  if (Array.isArray(data)) {
    return data.map(row => ({
      ...row,
      data: chunk(prepareData({ data: row.data, numColumns: col }), col)
    }));
  } else if (typeof data === 'object') {
    // support of v1
    return Object.keys(data).map(title => ({
      title,
      data: chunk(prepareData({ data: data[title], numColumns: col }), col)
    }));
  }
  return [];
};

export default class Grid extends Component {
  static propTypes = {
    itemsPerRow: PropTypes.number,
    numColumns: PropTypes.number,
    renderItem: PropTypes.func.isRequired,
    renderPlaceholder: PropTypes.func,
    data: PropTypes.arrayOf(PropTypes.any).isRequired,
    sections: PropTypes.bool,
    style: PropTypes.object
  };

  static defaultProps = {
    itemsPerRow: 3,
    renderPlaceholder: null,
    sections: false
  };

  constructor(props) {
    super(props);
    if (props.sections === true) {
      this.state = {
        data: prepareSectionedData(this.props)
      };
    } else {
      this.state = {
        data: prepareData(this.props)
      };
    }
  }

  static getDerivedStateFromProps(nextProps) {
    if (nextProps.sections === true) {
      return {
        data: prepareSectionedData(nextProps)
      };
    } else {
      return {
        data: prepareData(nextProps)
      };
    }
  }

  _renderPlaceholder = () => (
    <View style={{ width: width / this.props.itemsPerRow }} />
  );

  _renderItem = ({ item, index }) => {
    if (item) {
      if (typeof item === 'object' && item.__empty) {
        // render a placeholder
        if (this.props.renderPlaceholder) {
          return this.props.renderPlaceholder(index);
        }
        return this._renderPlaceholder(index);
      }
      return this.props.renderItem(item, index);
    }
  };

  _renderRow = ({ item }) => (
    <View style={styles.row}>
      {item.map((item, i) => {
        if (item) {
          return this.props.renderItem(item, i);
        }
        // render a placeholder
        if (this.props.renderPlaceholder) {
          return this.props.renderPlaceholder(i);
        }
        return this._renderPlaceholder(i);
      })}
    </View>
  );

  render() {
    // TODO: find a better way to filter props that we pass to ListView
    /* eslint-disable no-unused-vars */
    const {
      renderPlaceholder,
      renderItem,
      itemsPerRow,
      sections,
      style,
      data,
      numColumns, // To depricate itemsPerRow
      ...props
    } = this.props;
    /* eslint-enable no-unused-vars */

    return (
      <View style={styles.container}>
        {sections ? (
          <SectionList
            {...props}
            sections={this.state.data}
            numColumns={numColumns || itemsPerRow}
            style={[styles.list, style]}
            renderItem={this._renderRow}
          />
        ) : (
          <FlatList
            {...props}
            data={this.state.data}
            numColumns={numColumns || itemsPerRow}
            style={[styles.list, style]}
            renderItem={this._renderItem}
          />
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  list: {
    flex: 1
  },
  row: {
    justifyContent: 'space-around',
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'flex-start',
    flex: 1
  }
});


Examples/Simple/example.js
/**
 * @flow
 */

import React, { Component } from 'react';
import {
  StyleSheet,
  View,
} from 'react-native';

import Grid from 'react-native-grid-component';

export default class Simple extends Component {

  constructor(props) {
    super(props);

    this.state = {
      data: generateRandomColorsArray(21),
    };
  }

  _renderItem = (data, i) =>
    <View style={[{ backgroundColor: data }, styles.item]} key={i} />

  render() {
    return (
      <Grid
        style={styles.list}
        renderItem={this._renderItem}
        data={this.state.data}
        itemsPerRow={2}
        itemHasChanged={(d1, d2) => d1 !== d2}
        onEndReached={() =>
          this.setState({ data: [...this.state.data, ...generateRandomColorsArray(21)] })}
      />
    );
  }
}

const styles = StyleSheet.create({
  item: {
    flex: 1,
    height: 160,
    margin: 1,
  },
  list: {
    flex: 1,
  },
});

// Helper functions
// thanks materialuicolors.co
const colors = [
  '#F44336', '#E91E63', '#9C27B0', '#673AB7', '#3F51B5',
  '#2196F3', '#03A9F4', '#00BCD4', '#009688', '#4CAF50',
  '#8BC34A', '#CDDC39', '#FFEB3B', '#FFC107', '#FF9800',
  '#FF5722', '#795548', '#9E9E9E', '#607D8B',
];

function generateRandomColorsArray(length) {
  return Array.from(Array(length)).map(() => colors[Math.floor(Math.random() * colors.length)]);
}


Examples/Simple/ios/Simple/AppDelegate.m
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#import "AppDelegate.h"

#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  NSURL *jsCodeLocation = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index.ios" fallbackResource:nil];

  RCTRootView *rootView = [[RCTRootView alloc] initWithBundleURL:jsCodeLocation
                                                      moduleName:@"Simple"
                                               initialProperties:nil
                                                   launchOptions:launchOptions];
  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  return YES;
}

@end


Examples/Simple/android/app/src/main/java/com/simple/MainApplication.java
package com.simple;

import android.app.Application;
import android.util.Log;

import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.shell.MainReactPackage;
import com.facebook.soloader.SoLoader;

import java.util.Arrays;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
    @Override
    public boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
          new MainReactPackage()
      );
    }
  };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
  }
}


example/App.js
import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';

import Grid from 'react-native-grid-component';

const ITEMS_COUNT = 19;

export default class Simple extends Component {
  constructor(props) {
    super(props);

    this.state = {
      data: generateRandomColorsArray(ITEMS_COUNT),
      refreshing: false
    };
  }

  _renderItem = data => (
    <View style={[{ backgroundColor: data }, styles.item]} />
  );

  _renderPlaceholder = () => <View style={styles.item} />;

  render() {
    return (
      <Grid
        style={styles.list}
        renderItem={this._renderItem}
        renderPlaceholder={this._renderPlaceholder}
        data={this.state.data}
        numColumns={2}
        keyExtractor={(item, index) => index.toString()}
        refreshing={this.state.refreshing}
        onRefresh={() => {
          this.setState({
            data: generateRandomColorsArray(ITEMS_COUNT),
            refreshing: false
          });
        }}
        onEndReached={() => {
          this.setState(({ data }) => ({
            data: [...data, ...generateRandomColorsArray(ITEMS_COUNT)]
          }));
        }}
      />
    );
  }
}

const styles = StyleSheet.create({
  item: {
    flex: 1,
    height: 160,
    margin: 1
  },
  list: {
    flex: 1
  }
});

// Helper functions
// thanks materialuicolors.co
const colors = [
  '#F44336',
  '#E91E63',
  '#9C27B0',
  '#673AB7',
  '#3F51B5',
  '#2196F3',
  '#03A9F4',
  '#00BCD4',
  '#009688',
  '#4CAF50',
  '#8BC34A',
  '#CDDC39',
  '#FFEB3B',
  '#FFC107',
  '#FF9800',
  '#FF5722',
  '#795548',
  '#9E9E9E',
  '#607D8B'
];

function generateRandomColorsArray(length) {
  return Array.from(Array(length)).map(
    () => colors[Math.floor(Math.random() * colors.length)]
  );
}
